File: backend/src/main/java/med/doctor_connect/service/MedicineService.java
package med.doctor_connect.service;

import med.doctor_connect.dto.MedicineDto;
import org.springframework.data.domain.Page;

import java.util.List;
import java.util.UUID;

public interface MedicineService {

    MedicineDto createMedicine(MedicineDto medicineDto);

    MedicineDto getMedicineById(UUID id);

    List<MedicineDto> getAllActiveMedicines();

    Page<MedicineDto> searchMedicines(String name, int page, int limit);

    MedicineDto updateMedicine(UUID id, MedicineDto medicineDto);

    void deleteMedicine(UUID id);
}

File: frontend/doctor-connect-ui/src/app/services/medicine.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Medicine, PaginatedResponse } from '../models';

@Injectable({
  providedIn: 'root'
})
export class MedicineService {
  private readonly API_URL = 'http://localhost:8000/api/v1';

  constructor(private http: HttpClient) {}

  createMedicine(medicine: Medicine): Observable<Medicine> {
    return this.http.post<Medicine>(`${this.API_URL}/admin/medicines`, medicine);
  }

  getMedicineById(medicineId: string): Observable<Medicine> {
    return this.http.get<Medicine>(`${this.API_URL}/admin/medicines/${medicineId}`);
  }

  getAllActiveMedicines(): Observable<Medicine[]> {
    return this.http.get<Medicine[]>(`${this.API_URL}/admin/medicines/active`);
  }

  searchMedicines(name: string = '', page: number = 0, limit: number = 20): Observable<PaginatedResponse<Medicine>> {
    const params = new HttpParams()
      .set('name', name)
      .set('page', page.toString())
      .set('limit', limit.toString());

    return this.http.get<PaginatedResponse<Medicine>>(`${this.API_URL}/admin/medicines`, { params });
  }

  updateMedicine(medicineId: string, medicine: Medicine): Observable<Medicine> {
    return this.http.put<Medicine>(`${this.API_URL}/admin/medicines/${medicineId}`, medicine);
  }

  deleteMedicine(medicineId: string): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/admin/medicines/${medicineId}`);
  }
}
